contrib/jms/src/main/java/com/google/cloud/dataflow/contrib/jms/JmsIO.java

```diff
@@ -0,0 +1,532 @@
+             * @return the Bound
+             */
+            public Bound<T> connectionFactory(ConnectionFactory connectionFactory) {
+                return new Bound<T>(name, connectionFactory, queue, topic, coder, maxNumMessages);
+            }
+
+            /**
+             * Returns a transform that's like this one but that reads
+             * from the specified JMS queue.
+             *
+             * @param queue the JMS queue
+             * @return the Bound
+             */
+            public Bound<T> queue(String queue) {
+                return new Bound<T>(name, connectionFactory, queue, topic, coder, maxNumMessages);
+            }
+
+            /**
+             * Returns a transform that's like this one but that reads
+             * from the specified JMS topic.
+             *
+             * @param topic the JMQ topic
+             * @return the Bound
+             */
+            public Bound<T> topic(String topic) {
+                return new Bound<T>(name, connectionFactory, queue, topic, coder, maxNumMessages);
+            }
+
+            /**
+             * Returns a transform that's like this one but that uses the
+             * given {@link Coder} to decode each record into a value of
+             * type {@code X}.
+             *
+             * @param coder the coder to use
+             * @return the Bound
+             */
+            public <X> Bound<X> withCoder(Coder<X> coder) {
+                return new Bound<>(name, connectionFactory, queue, topic, coder, maxNumMessages);
+            }
+
+            /**
+             * Return a transform that's like this one but will only read up
+             * to the specified maximum number of messages from the JMS broker.
+             * The transform produces a <i>bounded</i> {@link PCollection}.
+             *
+             * @param maxNumMessages the max number messages limit to consume
+             * @return the Bound
+             */
+            public Bound<T> maxNumMessages(int maxNumMessages) {
+                return new Bound<T>(name, connectionFactory, queue, topic, coder, maxNumMessages);
+            }
+
+            @Override
+            public PCollection<T> apply(PInput input) {
+                if (connectionFactory == null) {
+                    throw new IllegalArgumentException("need to set connection factory " +
+                            "for a JmsIO.Read transform");
+                }
+                if (queue == null && topic == null) {
+                    throw new java.lang.IllegalStateException("need to set destination " +
+                            "(queue or topic) for a JmsIO.Read transform");
+                }
+
+                boolean boundedOutput = getMaxNumMessages() > 0;
+
+                if (boundedOutput) {
+                    return input.getPipeline().begin()
+                            .apply(Create.of((Void) null)).setCoder(VoidCoder.of())
```

Comment:
We recommend using [`Create#withCoder`](https://github.com/GoogleCloudPlatform/DataflowJavaSDK/blob/master/sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/Create.java#L70) instead of setting the coder on the output PCollection.

The reasoning is a little subtle -- any PTransform that changes the type of its input at runtime needs a strong handle on the TypeDescriptor (or at least the Coder) that a user may want to use. This is because the runner-specific implementation of `Create` may be a composite PTransform that needs the coder in `apply`. Note that the output PCollection is created during apply time, so naturally the output coder could not be set yet.